native/src/main/java/io/ballerina/lib/data/xmldata/utils/DataUtils.java

```diff
@@ -608,6 +609,44 @@ private static void processArray(Type childType, BMap<BString, Object> annotatio
                     StringUtils.fromString(getElementName(((RecordType) elementType).getAnnotations(), keyName)),
                     ValueCreator.createArrayValue(records.toArray(),
                             TypeCreator.createArrayType(Constants.JSON_ARRAY_TYPE)));
+        } else if (elementType.getTag() == TypeTags.UNION_TAG) {
+            Map<String, List<Object>> records = new HashMap<>();
+            for (int i = 0; i < arrayValue.getLength(); i++) {
+                Object element = arrayValue.get(i);
+                Type actualElementType = getTypeFromUnionType(elementType, element);
+                if (actualElementType.getTag() == TypeTags.RECORD_TYPE_TAG) {
```

Comment:
if the expected type is TypeTags.MAP_TYPE_TAG, this behaviour should be same ryt?